src/services/uniswap.js

```diff
@@ -0,0 +1,329 @@
+    if (!route) {
+      reportLog('Unable to find a possible route', null, 'error');
+    }
+  }
+  const trade = await getTrade(fromAsset.address, fromAssetQuantityBaseUnits, route);
+  const expectedOutput: BigNumber = getExpectedOutput(trade);
+  const toAssetDecimalsBN = new BigNumber(toAssetDecimals);
+  const expectedOutputWithSlippage = applyAllowedSlippage(expectedOutput, toAssetDecimalsBN);
+  const expectedOutputWithSlippageBaseUnits = convertToBaseUnits(toAssetDecimalsBN, expectedOutputWithSlippage);
+  const mappedPath: string[] = route.path.map(p => p.address);
+
+  return {
+    path: mappedPath,
+    expectedOutputBaseUnits: expectedOutputWithSlippageBaseUnits,
+  };
+};
+
+export const createUniswapOrder = async (
+  fromAsset: Asset,
+  toAsset: Asset,
+  quantity: number | string,
+  clientSendAddress: string,
+): Promise<Object> => {
+  if (!fromAsset || !toAsset) {
+    reportOrWarn('Invalid assets', null, 'error');
+  }
+
+  const decimalsBN = new BigNumber(fromAsset.decimals);
+  const quantityBN = new BigNumber(quantity);
+  const quantityBaseUnits = convertToBaseUnits(decimalsBN, quantityBN);
+
+  let txData = '';
+  let txValue = '0';
+  if (fromAsset.code !== 'ETH' && toAsset.code !== 'ETH') {
+    const { path, expectedOutputBaseUnits } = await getUniswapOrderData(
+      fromAsset,
+      toAsset,
+      quantityBaseUnits.toFixed(),
+      toAsset.decimals.toString(),
+    );
+    const deadline = getDeadline();
+
+    txData = swapExactTokensToTokens(
+      quantityBaseUnits.toFixed(),
+      expectedOutputBaseUnits.toFixed(),
+      path,
+      clientSendAddress,
+      deadline.toString(),
+    );
+  } else if (fromAsset.code === 'ETH' && toAsset.code !== 'ETH') {
+    txValue = quantityBaseUnits.toFixed();
+    const { path, expectedOutputBaseUnits } = await getUniswapOrderData(
+      WETH[chainId],
+      toAsset,
+      quantityBaseUnits.toFixed(),
+      toAsset.decimals.toString(),
+    );
+
+    const deadline = getDeadline();
+
+    txData = swapExactEthToTokens(
+      expectedOutputBaseUnits.toFixed(),
+      path,
+      clientSendAddress,
+      deadline.toString(),
+    );
+  } else if (fromAsset.code !== 'ETH' && toAsset.code === 'ETH') {
+    const { path, expectedOutputBaseUnits } = await getUniswapOrderData(
+      fromAsset,
+      WETH[chainId],
+      quantityBaseUnits.toFixed(),
+      toAsset.decimals.toString(),
+    );
+
+    const deadline = getDeadline();
+
+    txData = swapExactTokensToEth(
+      quantityBaseUnits.toFixed(),
+      expectedOutputBaseUnits.toFixed(),
+      path,
+      clientSendAddress,
+      deadline.toString(),
+    );
+  }
+
+  if (!txData) {
+    reportOrWarn('Unable to create order', null, 'error');
+  }
+
+  const txCount = await ethProvider.getTransactionCount(clientSendAddress);
+  const txObject = generateTxObject(
+    txCount.toString(),
+    ADDRESSES.router,
+    txValue,
+    txData,
```

Comment:
All good if `txData` is undefined? `generateTxObject ` seems to accept it as not nullable string.